src/sgw_filter.jl

```diff
@@ -0,0 +1,231 @@
+"""
+    struct WeightedSavitzkyGolayFilter{T<:RealQuantity} <: AbstractRadFIRFilter
+
+A [Weighted-Savitzky-Golay filter](https://doi.org/10.1021/acsmeasuresciau.3c00017).
+Working example:
+```julia
+using RadiationDetectorSignals
+using Unitful
+
+n = 600
+noise = 1.
+t = range(0u"μs", 20u"μs", 2*n)
+signal = vcat(zeros(n), 10*ones(n)) + (noise*rand(2*n) .- noise/2)
+wf = RDWaveform(t, signal)
+
+# define filter parameters and filter
+flt = WeightedSavitzkyGolayFilter(length=1u"μs", degree=3, weightType=2)
+
+# apply filter to signal
+wf_new = flt(wf)
+```
+
+
+Constructors:
+
+* ```$(FUNCTIONNAME)(; fields...)```
+
+Fields:
+
+$(TYPEDFIELDS)
+"""
+Base.@kwdef struct WeightedSavitzkyGolayFilter{T<:RealQuantity} <: AbstractRadFIRFilter
+    "total filter length"
+    length::T
+
+    "Polynomial degree"
+    degree::Int = 1
+
+    "weight function to use"
+    weightType::Int = 0
+
+    function WeightedSavitzkyGolayFilter(length::T, degree::Int, 
+        weightType::Int) where {T}
+
+        @assert weightType ∈ WEIGHTTYPES "unvalid weight function provided"
+
+        new{T}(length, degree, weightType)
+    end
+end
+
+export WeightedSavitzkyGolayFilter
+
+struct WeightedSavitzkyGolayFilterInstance{T<:RealQuantity} <: AbstractRadSigFilterInstance{NonlinearFiltering}
+    m::Int
+    degree::Int
+    weightType::Int
+    n_input::Int
+end
+
+const SGW_COEFFS = Dict(
+    0 => [1., 1., -1.],                  # normal Sawitzky-Golay filter
+    1 => [0.68096, 0.36358, -3.68528],   # GAUSS2
+    2 => [0.67574, 0.35440, -3.61580],   # HANN
+    3 => [0.63944, 0.28417, -5.508],     # HANNSQR
+    4 => [0.62303, 0.25310, -7.07317]    # HANNCUBE
+)
+
+const WEIGHTTYPES = (0, 1, 2, 3, 4)
+
+function fltinstance(flt::WeightedSavitzkyGolayFilter, fi::SamplingInfo{T}
+    ) where {T}
+
+    fltlen = round(Int, ustrip(NoUnits, flt.length / step(fi.axis)))
```

Comment:
The `ustrip(NoUnits, ...)` is used to really demand the `flt.length/step(fi.axis)` to have no units?

Reply:
It also converts the units of `flt.length` and `step(fi.axis)` to a common unit before computing the fraction